File: Simple_NN_library/C++/src/functions/loss_functions/regressive.cpp
#include <functions/loss_functions/regressive.h>

using namespace std;
using namespace functions;
using namespace loss_functions;
using namespace the_network;
using namespace further_maths;

string Regressive::get_func_name(float* func){
  string func_name="unknown";
  if(func==this.list_of_functions[0]){func_name="mean square error";}
  else if(func==this.list_of_functions[1]){func_name="absolute error";}
  else if(func==this.list_of_functions[2]){func_name="mean absolute percentage error";}
  else if(func==this.list_of_functions[3]){func_name="mean squared logarithmic error";}
  return func_name;
}
float* Regressive::get_func(string name){
  if(name=="mean square error"){return this.list_of_functions[0];}
  else if(name=="absolute error"){return this.list_of_functions[1];}
  else if(name=="mean absolute percentage error"){return this.list_of_functions[2];}
  else if(name=="mean squared logarithmic error"){return this.list_of_functions[3];}
}

float Regressive::mean_square_error(float output[], float expected_output[]){
  return mean(square_array(subtract_arrays(output, expected_output)));
}

float Regressive::absolute_error(float output[], float expected_output[]){
  return mean(array_abs(subtract_arrays(output, expected_output)));
}

float Regressive::mean_absolute_percentage_error(float output[], float expected_output[]){
  //return 100*mean(divide_arr(array_abs(subtract_arrays(expected_output, output)), array_abs()));
  float result[];
  for(int i=0; i<sizeof(output); i++){
    result[i]=abs((expected_output[i]-output[i])/expected_output[i])
  }
  return (100/sizeof(output))*sum(result);
}

float Regressive::mean_squared_logarithmic_error(float output[], float expected_output){
  return mean(square_array(subtract_arrays(array_log(output), array_log(expected_output))));
}

File: Simple_NN_library/python/functions/loss_functions/regressive.py
from further_maths import*

class Regressive:

    def get_func_name(self, func):
        self.list_of_functions=[self.mean_square_error, self.absolute_error, self.mean_absolute_percentage_error, self.mean_squared_logarithmic_error]
        func_name="unknown";
        if func==self.list_of_functions[0]:func_name="mean square error"
        elif func==self.list_of_functions[1]:func_name="absolute error"
        elif func==self.list_of_functions[2]:func_name="mean absolute percentage error"
        elif func==self.list_of_functions[3]:func_name="mean squared logarithmic error"
        return func_name

    def get_func(self, name):
        self.list_of_functions=[self.mean_square_error, self.absolute_error, self.mean_absolute_percentage_error, self.mean_squared_logarithmic_error]
        if name=="mean square error":return self.list_of_functions[0]
        elif name=="absolute error":return self.list_of_functions[1]
        elif name=="mean absolute percentage error":return self.list_of_functions[2]
        elif name=="mean squared logarithmic error":return self.list_of_functions[3]

    def mean_square_error(self, output, expected_output):
        return mean(square_array(subtract_arrays(output, expected_output)))

    def absolute_error(self, output, expected_output):
        return mean(array_abs(subtract_arrays(output, expected_output)))

    def mean_absolute_percentage_error(self, output, expected_output):
        result=[]
        for i in range(len(output)):
            result.append(abs((expected_output[i]-output[i])/expected_output[i]))
        return (100/len(output))*sum(result)

    def mean_squared_logarithmic_error(self, output, expected_output):
        return mean(square_array(subtract_arrays(array_log(output), array_log(expected_output))))
